### ISSUE TITLE
ClassCastException when using Timestamp within DataSerializable 

### ISSUE THREAD

**Issue by GhostInAMachine** [hazelcast/hazelcast#6759]
Running the code below results in ClassCastException when data is being read during deserialization. Checked with 3.6-EA.

    public class TimestampIDS {
    
        public static void main(String[] args) {
            HazelcastInstance hazelcastInstance = Hazelcast.newHazelcastInstance(new Config());
            IMap<Integer, Container> timestamps = hazelcastInstance.getMap("timestamps");
            timestamps.put(1, new Container(new Timestamp(1337)));
            Container c = timestamps.get(1);
            System.out.println(c.getTimestamp());
        }
    
        public static class Container implements DataSerializable {
    
            public Container() {
            }
    
            public Container(Timestamp timestamp) {
                this.timestamp = timestamp;
            }
    
            private Timestamp timestamp;
    
            public Timestamp getTimestamp() {
                return timestamp;
            }

            public void writeData(ObjectDataOutput out) throws IOException {
                out.writeObject(timestamp);
            }
    
            public void readData(ObjectDataInput in) throws IOException {
                timestamp = in.readObject();
            }
        }
    }

**Comment by pveentjer** [comment]
I just tried it with master and it runs fine. From your example it isn't clear which package Timestamp comes from; so I used 
>>> import java.sql.Timestamp;.

**Comment by pveentjer** [comment]
Ahh.. 

Apparently it is resolved in 3.6 master. when I switch to 3.5.x I get the following exception:

```
Exception in thread "main" com.hazelcast.nio.serialization.HazelcastSerializationException: Problem while reading DataSerializable, namespace: 0, id: 0, class: 'com.hazelcast.Main$Container', exception: java.util.Date cannot be cast to java.sql.Timestamp
	at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:124)
	at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:39)
	at com.hazelcast.nio.serialization.StreamSerializerAdapter.read(StreamSerializerAdapter.java:41)
	at com.hazelcast.nio.serialization.SerializationServiceImpl.toObject(SerializationServiceImpl.java:276)
	at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:200)
	at com.hazelcast.map.impl.MapServiceContextImpl.toObject(MapServiceContextImpl.java:311)
	at com.hazelcast.map.impl.proxy.MapProxySupport.toObject(MapProxySupport.java:1135)
	at com.hazelcast.map.impl.proxy.MapProxyImpl.get(MapProxyImpl.java:82)
	at com.hazelcast.Main.main(Main.java:21)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
Caused by: java.lang.ClassCastException: java.util.Date cannot be cast to java.sql.Timestamp
	at com.hazelcast.Main$Container.readData(Main.java:45)
	at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:111)
	... 13 more
```

@asimarslan I believe you have been working in this area. Do you know exactly which PR fixed this bug and if it is going to be backported?

**Comment by asimarslan** [comment]
@pveentjer the related PR is https://github.com/hazelcast/hazelcast/pull/6735
But it very difficult to backport this as it breaks binary compatibility of serialization.
The simplest solution is move Date.class to Constant serializers

**Comment by pveentjer** [comment]
@asimarslan what should we do with this ticket? I'm fine if it can't be backported and since it is already solved in master, perhaps we should close this ticket?

**Comment by asimarslan** [comment]
This issue is clone of https://github.com/hazelcast/hazelcast/issues/6455 and it is fixed in 3.6 so let's just close this one.
